Code/Framework/AzCore/AzCore/Debug/TraceReflection.cpp

```diff
@@ -48,63 +94,197 @@ namespace AZ
         }
 
         inline bool TraceMessageBusHandler::OnWarning(const char* window, const char* message)
         {
-            return CallResultReturn(false, FN_OnWarning, window, message);
+            QueueMessageCall(
+                [this, windowString = AZStd::string(window), messageString = AZStd::string(message)]()
+            {
+                return Call(FN_OnWarning, windowString.c_str(), messageString.c_str());
+            });
+            return m_defaultReturnValue;
         }
 
         inline bool TraceMessageBusHandler::OnException(const char* message)
         {
-            return CallResultReturn(false, FN_OnException, message);
+            QueueMessageCall(
+                [this, messageString = AZStd::string(message)]()
+            {
+                return Call(FN_OnException, messageString.c_str());
+            });
+            return m_defaultReturnValue;
         }
 
         inline bool TraceMessageBusHandler::OnPrintf(const char* window, const char* message)
         {
-            return CallResultReturn(false, FN_OnPrintf, window, message);
+            QueueMessageCall(
+                [this, windowString = AZStd::string(window), messageString = AZStd::string(message)]()
+            {
+                return Call(FN_OnPrintf, windowString.c_str(), messageString.c_str());
+            });
+            return m_defaultReturnValue;
         }
 
         inline bool TraceMessageBusHandler::OnOutput(const char* window, const char* message)
         {
-            return CallResultReturn(false, FN_OnOutput, window, message);
+            QueueMessageCall(
+                [this, windowString = AZStd::string(window), messageString = AZStd::string(message)]()
+            {
+                return Call(FN_OnOutput, windowString.c_str(), messageString.c_str());
+            });
+            return m_defaultReturnValue;
+        }
+
+        void TraceMessageBusHandler::OnTick(
+            [[maybe_unused]] float deltaTime,
+            [[maybe_unused]] AZ::ScriptTimePoint time)
+        {
+            FlushMessageCalls();
         }
 
+        int TraceMessageBusHandler::GetTickOrder()
+        {
+            return AZ::TICK_LAST;
+        }
+
+        void TraceMessageBusHandler::QueueMessageCall(AZStd::function<void()> messageCall)
+        {
+            AZStd::lock_guard<decltype(m_lock)> lock(m_lock);
+            m_messageCalls.push_back(messageCall);
+        }
+
+        void TraceMessageBusHandler::FlushMessageCalls()
+        {
+            AZStd::list<AZStd::function<void()>> messageCalls;
+            {
+                AZStd::lock_guard<decltype(m_lock)> lock(m_lock);
+                if (m_messageCalls.empty())
+                {
+                    return;
+                }
+                messageCalls = AZStd::move(m_messageCalls); // Move calls to release the lock as soon as possible
+                m_messageCalls = AZStd::list<AZStd::function<void()>>(); // Reset list
```

Comment:
you can just call ```messageCalls .swap(m_messageCalls);``` instead of the move + reconstruct.

Reply:
it should also cleanly handle the empty case, so you could remove the if empty() as well.

Reply:
Good call!